File: day14/sol.py
import copy

from timing_util import Timing


def get_data(data: str) -> list[list[str]]:
    return [list(line) for line in data.split("\n")]


def step_north(state: list[list[str]]):
    has_changed = True
    while has_changed:
        has_changed = False
        for y, row in enumerate(state[1:], 1):
            for x, val in enumerate(row):
                if val == "O" and state[y - 1][x] == ".":
                    i = y - 1
                    while i > 0 and state[i - 1][x] == ".":
                        i -= 1

                    state[i][x] = "O"
                    state[y][x] = "."
                    has_changed = True


def score(state: list[list[str]]) -> int:
    s = 0
    for score, row in enumerate(reversed(state), 1):
        for val in row:
            if val == "O":
                s += score
    return s


def part1(data: list[list[str]]) -> int:
    step_north(state := copy.deepcopy(data))
    return score(state)


def step_west(state: list[list[str]]):
    has_changed = True
    while has_changed:
        has_changed = False
        for y, row in enumerate(state):
            for x, val in reversed(list(enumerate(row[1:], 1))):
                if val == "O" and state[y][x - 1] == ".":
                    i = x - 1
                    while i > 0 and state[y][i - 1] == ".":
                        i -= 1

                    state[y][x - 1] = "O"
                    state[y][x] = "."
                    has_changed = True


def step_south(state: list[list[str]]):
    has_changed = True
    while has_changed:
        has_changed = False
        for y, row in reversed(list(enumerate(state[:-1]))):
            for x, val in enumerate(row):
                if val == "O" and state[y + 1][x] == ".":
                    i = y + 1
                    while i < len(state) - 1 and state[i + 1][x] == ".":
                        i += 1

                    state[i][x] = "O"
                    state[y][x] = "."
                    has_changed = True


def step_east(state: list[list[str]]):
    has_changed = True
    while has_changed:
        has_changed = False
        for y, row in enumerate(state):
            for x, val in enumerate(row[:-1]):
                if val == "O" and state[y][x + 1] == ".":
                    i = x + 1
                    while i < len(row) - 1 and state[y][i + 1] == ".":
                        i += 1

                    state[y][i] = "O"
                    state[y][x] = "."
                    has_changed = True


def cycle(state: list[list[str]]):
    step_north(state)
    step_west(state)
    step_south(state)
    step_east(state)


def to_hashable(data: list[list[str]]) -> str:
    return "".join(["".join(row) for row in data])


def part2(data: list[list[str]], n_cycles: int = 1_000_000_000) -> int:
    state_to_iteration = {}
    iteration_to_state = []

    i, state = 0, data
    while (hashed := to_hashable(state)) not in state_to_iteration:
        state_to_iteration[hashed] = i
        iteration_to_state.append(copy.deepcopy(state))

        cycle(state)
        i += 1

    first_reached = state_to_iteration[to_hashable(state)]
    cycles_remaining = n_cycles - first_reached
    cycles_remaining %= i - first_reached

    state = iteration_to_state[first_reached + cycles_remaining]
    return score(state)


if __name__ == "__main__":
    with Timing("Elapsed time: "):
        with open("inputs/day14.txt") as in_file:
            data = get_data(in_file.read().strip())

        print(f"part1: {part1(data)}")
        print(f"part2: {part2(data)}")

File: day14/sol.go
package main

import (
	"fmt"
	"os"
	"strings"
	"time"
)

func ReadFile(filename string) string {
	bytes, err := os.ReadFile(filename)
	if err != nil {
		panic(err)
	}
	return strings.TrimSpace(string(bytes))
}

func ParseInput(data string) [][]rune {
	grid := strings.Split(data, "\n")
	result := make([][]rune, len(grid))
	for i, line := range grid {
		result[i] = []rune(line)
	}
	return result
}

func StepNorth(state [][]rune) {
	has_changed := true
	for has_changed {
		has_changed = false
		for y := 1; y < len(state); y++ {
			for x, val := range state[y] {
				if val == 'O' && state[y-1][x] == '.' {
					i := y - 1
					for i >= 1 && state[i-1][x] == '.' {
						i--
					}

					state[i][x] = 'O'
					state[y][x] = '.'
					has_changed = true
				}
			}
		}
	}
}

func Score(state [][]rune) int {
	s := 0
	for i, row := range state {
		for _, val := range row {
			if val == 'O' {
				s += len(state) - i
			}
		}
	}
	return s
}

func clone(data [][]rune) [][]rune {
	result := make([][]rune, len(data))
	for i, row := range data {
		result[i] = make([]rune, len(row))
		copy(result[i], row)
	}
	return result
}

func Part1(data [][]rune) int {
	return Score(clone(data))
}

func StepWest(state [][]rune) {
    has_changed := true
	for has_changed {
		has_changed = false
		for _, row := range state {
			for x := 1; x < len(row); x++ {
				val := row[x]
				if val == 'O' && row[x-1] == '.' {
					i := x - 1
					for i >= 1 && row[i-1] == '.' {
						i--
					}

					row[i] = 'O'
					row[x] = '.'
					has_changed = true
				}
			}
		}
	}
}

func StepSouth(state [][]rune) {
    has_changed := true
	for has_changed {
        has_changed = false
		for y := len(state) - 2; y >= 0; y-- {
            for x, val := range state[y] {
                if val == 'O' && state[y+1][x] == '.' {
                    i := y + 1
					for i < len(state)-1 && state[i+1][x] == '.' {
                        i++
					}
                    
					state[i][x] = 'O'
					state[y][x] = '.'
					has_changed = true
				}
			}
		}
	}
}

func StepEast(state [][]rune) {
    has_changed := true
    for has_changed {
        has_changed = false
        for _, row := range state {
            for x := len(row) - 2; x >= 0; x-- {
                val := row[x]
                if val == 'O' && row[x+1] == '.' {
                    i := x + 1
                    for i < len(row)-1 && row[i+1] == '.' {
                        i++
                    }

                    row[i] = 'O'
                    row[x] = '.'
                    has_changed = true
                }
            }
        }
    }
}

func Cycle(state [][]rune) {
    StepNorth(state)
	StepWest(state)
	StepSouth(state)
	StepEast(state)
}

func ToString(data [][]rune) string {
	s := make([]rune, (len(data)+1)*len(data[0]))
	i := 0
	for _, row := range data {
		copy(s[i:], row)
		i += len(row)
        // make it printable as well :)
		s[i] = '\n'
		i++
	}
	return string(s)
}

func Part2(data [][]rune, n_cycles int) int {
	state_to_iteration := make(map[string]int)
	iteration_to_state := make([][][]rune, 0)

	state := data
	for i := 0; ; i++ {
		str := ToString(state)
		if _, ok := state_to_iteration[str]; ok {
			break
		}
		state_to_iteration[str] = i
		iteration_to_state = append(iteration_to_state, clone(state))

		Cycle(state)
	}

	first_reached := state_to_iteration[ToString(data)]
	cycles_remaining := n_cycles - first_reached
	cycles_remaining %= len(iteration_to_state) - first_reached

	state = iteration_to_state[first_reached+cycles_remaining]
	return Score(state)
}

func PrintTime(start time.Time, msg string) {
	dur := float64(time.Since(start).Nanoseconds())
	switch {
	case dur < 1e3:
		fmt.Printf("%v %.3fns\n", msg, dur)
	case dur < 1e6:
		fmt.Printf("%v %.3fµs\n", msg, dur/1e3)
	case dur < 1e9:
		fmt.Printf("%v %.3fms\n", msg, dur/1e6)
	default:
		fmt.Printf("%v %.3fs\n", msg, dur/1e9)
	}
}

func main() {
	start := time.Now()
	defer PrintTime(start, "Elapsed time:")

	text := ReadFile("inputs/day14.txt")
	data := ParseInput(text)

	fmt.Printf("Part1: %d\n", Part1(data))
	fmt.Printf("Part2: %d\n", Part2(data, 1_000_000_000))
}
